packages/core/mesh/edge-client/src/edge-client.ts

```diff
@@ -0,0 +1,204 @@
+  addListener(listener: MessageListener): () => void;
+  send(message: Message): Promise<void>;
+}
+
+export type MessengerConfig = {
+  socketEndpoint: string;
+  timeout?: number;
+  protocol?: Protocol;
+};
+
+/**
+ * Messenger client.
+ */
+export class EdgeClient extends Resource implements EdgeConnection {
+  public reconnect = new Event();
+  private readonly _persistentLifecycle = new PersistentLifecycle({
+    start: async () => this._openWebSocket(),
+    stop: async () => this._closeWebSocket(),
+    onRestart: async () => this.reconnect.emit(),
+  });
+
+  private readonly _listeners = new Set<MessageListener>();
+  private readonly _protocol: Protocol;
+  private _ready = new Trigger();
+  private _ws?: WebSocket = undefined;
+  private _keepaliveCtx?: Context = undefined;
+
+  constructor(
+    private _identityKey: string,
+    private _peerKey: string,
+    private readonly _config: MessengerConfig,
+  ) {
+    super();
+    this._protocol = this._config.protocol ?? protocol;
+  }
+
+  // TODO(burdon): Attach logging.
+  public get info() {
+    return {
+      open: this.isOpen,
+      identity: this._identityKey,
+      device: this._peerKey,
+    };
+  }
+
+  get identityKey() {
+    return this._identityKey;
+  }
+
+  get peerKey() {
+    return this._peerKey;
+  }
+
+  public get isOpen() {
+    return this._lifecycleState === LifecycleState.OPEN;
+  }
+
+  setIdentity({ peerKey, identityKey }: { peerKey: string; identityKey: string }) {
+    this._peerKey = peerKey;
+    this._identityKey = identityKey;
+    this._persistentLifecycle.scheduleRestart();
+  }
+
+  public addListener(listener: MessageListener): () => void {
+    this._listeners.add(listener);
+    return () => this._listeners.delete(listener);
+  }
+
+  /**
+   * Open connection to messaging service.
+   */
+  protected override async _open() {
+    log('opening...', { info: this.info });
+    await this._persistentLifecycle.open();
+  }
+
+  /**
+   * Close connection and free resources.
+   */
+  protected override async _close() {
+    log('closing...', { peerKey: this._peerKey });
+    await this._persistentLifecycle.close();
+  }
+
+  private async _openWebSocket() {
+    const url = new URL(`/ws/${this._identityKey}/${this._peerKey}`, this._config.socketEndpoint);
+    this._ws = new WebSocket(url);
+
+    this._ws.onopen = () => {
+      log('opened', this.info);
+      this._ready.wake();
+    };
+    this._ws.onclose = () => {
+      log('closed', this.info);
+      this._persistentLifecycle.scheduleRestart();
+    };
+    this._ws.onerror = (event) => {
+      log.warn('EdgeClient socket error', { error: event.error, info: event.message });
+      this._persistentLifecycle.scheduleRestart();
+    };
+    /**
+     * https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent/data
+     */
+    this._ws.onmessage = async (event) => {
+      if (String(event.data) === '__pong__') {
```

Comment:
is `event.data` different when it's the pong auto-reply?

can we avoid converting every buffer sent to a string?